src/tokenserver/auth/browserid.rs

```diff
@@ -0,0 +1,469 @@
+#[derive(Serialize)]
+struct VerifyRequest {
+    assertion: String,
+    audience: String,
+    #[serde(rename(serialize = "trustedIssuers"))]
+    trusted_issuers: [String; 1],
+}
+
+/// The response returned by the FxA BrowserID verifier for a token verification request.
+#[derive(Deserialize, Serialize)]
+#[serde(tag = "status", rename_all = "lowercase")]
+enum VerifyResponse {
+    Okay {
+        email: String,
+        #[serde(rename = "idpClaims")]
+        idp_claims: Option<IdpClaims>,
+        issuer: String,
+    },
+    Failure {
+        reason: Option<String>,
+    },
+}
+
+/// The claims extracted from a valid BrowserID assertion.
+#[derive(Deserialize, Serialize)]
+struct IdpClaims {
+    #[serde(rename = "fxa-deviceId")]
+    pub device_id: Option<String>,
+    #[serde(
+        default,
+        rename = "fxa-generation",
+        deserialize_with = "strict_deserialize"
+    )]
+    generation: Option<Option<i64>>,
+    #[serde(
+        default,
+        rename = "fxa-keysChangedAt",
+        deserialize_with = "strict_deserialize"
+    )]
+    keys_changed_at: Option<Option<i64>>,
+    #[serde(
+        default,
+        rename = "fxa-tokenVerified",
+        deserialize_with = "strict_deserialize"
+    )]
+    token_verified: Option<Option<bool>>,
+}
+
+impl IdpClaims {
+    fn generation(&self) -> Result<Option<i64>, TokenserverError> {
+        match self.generation {
+            // If the fxa-generation claim is present, return its value. If it's missing, return None.
+            Some(Some(_)) | None => Ok(self.generation.flatten()),
+            // If the fxa-generation claim is null, return an error.
+            Some(None) => Err(TokenserverError::invalid_generation()),
+        }
+    }
+
+    fn keys_changed_at(&self) -> Result<Option<i64>, TokenserverError> {
+        match self.keys_changed_at {
+            // If the fxa-keysChangedAt claim is present, return its value. If it's missing, return None.
+            Some(Some(_)) | None => Ok(self.keys_changed_at.flatten()),
+            // If the fxa-keysChangedAt claim is null, return an error.
+            Some(None) => Err(TokenserverError {
+                description: "invalid keysChangedAt",
+                status: "invalid-credentials",
+                location: ErrorLocation::Body,
+                ..Default::default()
+            }),
+        }
+    }
+
+    fn token_verified(&self) -> bool {
+        match self.token_verified {
+            // If the fxa-tokenVerified claim is true or missing, return true.
+            Some(Some(true)) | None => true,
+            // If the fxa-tokenVerified claim is false or null, return false.
+            Some(Some(false)) | Some(None) => false,
+        }
+    }
+}
+
+// Approach inspired by: https://github.com/serde-rs/serde/issues/984#issuecomment-314143738
+fn strict_deserialize<'de, T, D>(deserializer: D) -> Result<Option<T>, D::Error>
```

Comment:
I use a custom deserializer here to differentiate between missing fields and null fields, since these cases are handled differently.

Reply:
These fields aren't mandatory, and I think older clients may not send them. The legacy Tokenserver is built to handle situations where the fields aren't sent: https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/views.py#L355-L361

Reply:
Yup, that's it. I'll definitely add some more comments to make it more obvious whats going on 👍🏻